src/Libraries/Nop.Services/Helpers/UserAgentHelper.cs

```diff
@@ -49,7 +54,33 @@ public virtual bool IsSearchEngine()
                     return false;
 
                 var userAgent = _httpContext.Request.UserAgent;
-                return _browscapXmlHelper.IsCrawler(userAgent);
+                if (String.IsNullOrEmpty(userAgent))
+                    return false;
+
+                // check in already found cache
+                bool result;
+                if (_foundSearchEnginesCache.TryGetValue(userAgent, out result))
+                    return true;
+
+                // check cookie
+                var cookie = _httpContext.Request.Cookies[IS_SEARCH_ENGINE_REQUEST_COOKIE_KEY];
+                if (cookie != null)
+                    return cookie.Value == "1";
+
+                // check browscap
+                result = _browscapXmlHelper.IsCrawler(userAgent);
+
+                // if IS search engine, save to cache
+                if (result)
+                    _foundSearchEnginesCache.TryAdd(userAgent, true);

```

Comment:
Yes good point, there should be a ceiling and honestly a mechanism to maintain itself, i.e ttl or sort strategy.

Could use memory cache manage for inherent ttl expiration?